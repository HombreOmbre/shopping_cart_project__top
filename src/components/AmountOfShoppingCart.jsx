import { useContext, useMemo } from "react";
import { ShoppingCartContext } from "../context/ShoppingCartContext.jsx";
import styles from "../styles/AmountOfShoppingCart.module.scss";

export const AmountOfShoppingCart = () => {
    const { shoppingCart } = useContext(ShoppingCartContext);
    const amount = useMemo(() => {
        return shoppingCart.reduce((allPrice, product) => {
            return allPrice + product.getProductPrice();
        }, 0);
    },[shoppingCart]);

    return (
        <div className={styles.container}>
            <div className={styles.amountTitle}>
                Amount:
            </div>
            <div className={styles.amountValue}>
                ${amount.toFixed(2)}
            </div>
        </div>
    );
}