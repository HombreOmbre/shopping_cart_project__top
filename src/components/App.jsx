import { useProducts } from "../customHooks/useProducts.jsx";

export const App = () => {
  const { products, error, loading } = useProducts();

  if (error) return <div>{error}</div>;
  if (loading) return <div>Loading...</div>;

  return (
      <ul>
      {
        products.length > 0 ?
             products.map((product) => {
               return (
                   <li key={product.id}>{product.title}</li>
               );
             }) : null
      }
      </ul>
  );
}
