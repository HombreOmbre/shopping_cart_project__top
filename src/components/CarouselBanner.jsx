import Carousel from "react-bootstrap/Carousel";
import 'bootstrap/dist/css/bootstrap.css';
import { useState } from "react";
import { Link } from "react-router-dom";
import { bannersData } from "../assets/banners.js";
import styles from "../styles/CarouselBanner.module.scss";

export const CarouselBanner = () => {
    const [index, setIndex] = useState(0);

    const handleSelect = (selectedIndex) => {
        setIndex(selectedIndex);
    }

    return (
        <div className={styles.mainContainer}>
            <Carousel activeIndex={index} onSelect={handleSelect} fade controls={false} indicators={false}  interval={1500} className={styles.container}>
                {
                    bannersData.map(item => (
                        <Carousel.Item key={item.id}>
                            <Link to="/">
                                <img src={item.img} alt={item.alt} className={styles.img} />
                            </Link>
                        </Carousel.Item>
                    ))
                }
            </Carousel>
        </div>
    );
}
