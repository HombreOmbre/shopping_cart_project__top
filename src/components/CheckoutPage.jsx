import { useContext } from "react";
import { ShoppingCartContext } from "../context/ShoppingCartContext.jsx";
import styles from "../styles/CheckoutPage.module.scss";
import { ShoppingCartItem } from "./ShoppingCartItem.jsx";

export const CheckoutPage = () => {
    const { shoppingCart } = useContext(ShoppingCartContext);
    return (
        <div className={styles.container}>

            {
                shoppingCart.length > 0 ? (
                    <>
                        <h2 className={styles.titleHeader}>Shopping cart</h2>
                        {
                            shoppingCart.map((product) => (
                                <ShoppingCartItem
                                    key={product.getId()}
                                    product={product}
                                />
                            ))
                        }

                    </>
                ) :
                <h2 className={styles.infoTxt}>There is no products in the shopping cart</h2>
            }
        </div>
    );
}