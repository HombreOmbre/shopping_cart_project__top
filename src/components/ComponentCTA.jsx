import { useRef, useState } from "react";
import styles from "../styles/ComponentCTA.module.scss";

export const ComponentCTA = () => {
    const emailInp = useRef(null);
    const [send, setSend] = useState(false);
    const [isValid, setIsValid] = useState(false);

    const submitHandler = (e) => {
        e.preventDefault();

        const regex = new RegExp(/[a-zA-Z\d]+@[a-zA-Z\d]+\.[a-zA-Z\d]{2,}/);

        if (emailInp.current.value.length === 0) {
            setIsValid(true);
            return;
        }

        if (!regex.test(emailInp.current.value.trim())) {
            setIsValid(true);
            return;
        }

        setSend(true);
        setIsValid(false);
    }

    return (
        <div className={styles.container}>
            <h2 className={styles.title}>Stay in touch</h2>
            <div className={styles.txtCTA}>
                Lorem ipsum dolor, sit amet consectetur adipisicing elit. Earum id, iure consectetur et error hic!
            </div>
            {!send ?
                <form className={styles.form}>
                    <input type="email" defaultValue='' ref={emailInp} className={styles.email} placeholder="Enter your email"/>
                    {
                        isValid &&
                        <span className={styles.errorEmail}>Email is invalid. Please try example@example.com</span>
                    }
                    <button type="submit" className={styles.submit} onClick={submitHandler}>Subscribe</button>
                </form> :
                <div className={styles.subscribe}>Thank You for subscribe!</div>
            }
        </div>
    );
}