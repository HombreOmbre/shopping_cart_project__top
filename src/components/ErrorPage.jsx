import styles from "../styles/ErrorPage.module.scss";

export const ErrorPage = () => {
    return (
        <div className={styles.errorPage}>
            <h1 className={styles.header}>Ooops!</h1>
            <p className={styles.errorTxt}>Something gone wrong...</p>
        </div>
    )
}