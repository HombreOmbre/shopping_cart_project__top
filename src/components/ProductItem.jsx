import { Link } from "react-router-dom";
import PropTypes from "prop-types";
import styles from "../styles/ProductItem.module.scss";

export const ProductItem = ({ product }) => {
    let color = styles.greenRate;
    switch (true) {
        case (product.rating.rate <= 2):
            color = styles.redRate;
            break;
        case (product.rating.rate > 2 && product.rating.rate <= 3.9):
            color = styles.orangeRate;
            break;
        case (product.rating.rate > 4):
            color = styles.greenRate;
        break;
    }

    return (
        <Link className={styles.linkContainer} to={`${product.id}`}>
            <div className={styles.imageContainer}>
                <img className={styles.cartImg} src={product.image} alt={product.title}/>
            </div>
            <div className={styles.details}>
                <h6 className={styles.title}>{product.title}</h6>
                <p className={styles.rate}>Rate:
                    <span className={color}> {product.rating.rate} </span>
                    ({product.rating.count})
                </p>
                <p className={styles.price}>${product.price.toFixed(2)}</p>
                <span className={styles.delivery}>Free delivery</span>
            </div>
        </Link>
    );
}

ProductItem.propTypes = {
    product: PropTypes.shape({
        id: PropTypes.number.isRequired,
        title: PropTypes.string.isRequired,
        price: PropTypes.number.isRequired,
        description: PropTypes.string.isRequired,
        category: PropTypes.string.isRequired,
        image: PropTypes.string.isRequired,
        rating: PropTypes.shape({
            rate: PropTypes.number.isRequired,
            count: PropTypes.number.isRequired,
        })
    })
}