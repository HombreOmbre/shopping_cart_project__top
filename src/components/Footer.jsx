import { ComponentCTA } from "./ComponentCTA.jsx";
import { MapSiteItem } from "./MapSiteItem.jsx";
import { mapSites } from "../assets/mapSites.js";
import styles from "../styles/Footer.module.scss";

export const Footer = () => {
    return (
        <footer className={styles.footer}>
            <div className={styles.footerContainer}>
                <div className={styles.footerContent}>
                    <div className={styles.title}>
                        SHOPPEROOO
                    </div>
                    <div className={styles.desc}>
                        Lorem ipsum dolor, sit amet consectetur adipisicing elit. Incidunt consequuntur amet culpa cum itaque neque.
                    </div>
                </div>
                <div className={styles.footerMapSite}>
                    {
                        mapSites.map((map, index) => (
                            <MapSiteItem key={index} title={map.title} linksToSite={map.links} />
                        ))
                    }
                    <ComponentCTA />
                </div>
                <div className={styles.footerRights}>
                    Copyright Â© HombreOmbre {new Date().getFullYear()}. All rights reserved.
                </div>
            </div>
        </footer>
    );
}