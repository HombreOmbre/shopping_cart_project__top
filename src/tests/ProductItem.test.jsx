import { render, screen } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { describe, it, expect } from 'vitest';
import { ProductItem } from '../components/ProductItem';
import '@testing-library/jest-dom';

const mockProduct = {
    id: 1,
    title: 'Test Product',
    price: 99.99,
    description: 'This is a test product',
    category: 'electronics',
    image: 'https://via.placeholder.com/150',
    rating: {
        rate: 4.5,
        count: 100,
    },
};

const renderComponent = (product = mockProduct) => {
    return render(
        <BrowserRouter>
            <ProductItem product={product} />
        </BrowserRouter>
    );
};

describe('ProductItem Component', () => {
    it('renders the product details correctly', () => {
        renderComponent();

        expect(screen.getByText(mockProduct.title)).toBeInTheDocument();

        expect(screen.getByText(`$${mockProduct.price}`)).toBeInTheDocument();

        expect(screen.getByText(`${mockProduct.rating.rate}`)).toBeInTheDocument();

        const productImg = screen.getByAltText(mockProduct.title);
        expect(productImg).toBeInTheDocument();
        expect(productImg).toHaveAttribute('src', mockProduct.image);

        expect(screen.getByText(/free delivery/i)).toBeInTheDocument();
    });

    it('applies correct color class based on rating', () => {
        let product = { ...mockProduct, rating: { rate: 4.5, count: 100 } };
        renderComponent(product);
        expect(screen.getByText(`${product.rating.rate}`)).toHaveClass(/greenRate/);

        product = { ...mockProduct, rating: { rate: 3.5, count: 50 } };
        renderComponent(product);
        expect(screen.getByText(`${product.rating.rate}`)).toHaveClass(/orangeRate/);

        product = { ...mockProduct, rating: { rate: 1.5, count: 20 } };
        renderComponent(product);
        expect(screen.getByText(`${product.rating.rate}`)).toHaveClass(/redRate/);
    });

    it('renders the correct link to the product details', () => {
        renderComponent();

        const productLink = screen.getByRole('link');
        expect(productLink).toHaveAttribute('href', `/${mockProduct.id}`);
    });

    it('matches snapshot', () => {
        const { asFragment } = renderComponent();
        expect(asFragment()).toMatchSnapshot();
    });
});
