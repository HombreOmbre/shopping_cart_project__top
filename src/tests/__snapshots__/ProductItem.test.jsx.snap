// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProductItem Component > matches snapshot 1`] = `
<DocumentFragment>
  <a
    class="_linkContainer_cca2e2"
    href="/1"
  >
    <div
      class="_imageContainer_cca2e2"
    >
      <img
        alt="Test Product"
        class="_cartImg_cca2e2"
        src="https://via.placeholder.com/150"
      />
    </div>
    <div
      class="_details_cca2e2"
    >
      <h6
        class="_title_cca2e2"
      >
        Test Product
      </h6>
      <p
        class="_rate_cca2e2"
      >
        Rate:
        <span
          class="_greenRate_cca2e2"
        >
           4.5 
        </span>
        (100)
      </p>
      <p
        class="_price_cca2e2"
      >
        $99.99
      </p>
      <span
        class="_delivery_cca2e2"
      >
        Free delivery
      </span>
    </div>
  </a>
</DocumentFragment>
`;
